imap_processing/hi/l1a/histogram.py

```diff
@@ -0,0 +1,155 @@
+"""Unpack IMAP-Hi histogram data."""
+
+import dataclasses
+
+import numpy as np
+import xarray as xr
+from space_packet_parser.parser import Packet
+
+from imap_processing.cdf.global_attrs import ConstantCoordinates
+from imap_processing.cdf.utils import calc_start_time
+from imap_processing.hi import hi_cdf_attrs
+
+# TODO: Verify that these names are OK for counter variables in the CDF
+# define the names of the 24 counter arrays
+# contained in the histogram packet
+QUALIFIED_COUNTERS = (
+    "qual_ab",
+    "qual_c1c2",
+    "qual_ac1",
+    "qual_bc1",
+    "qual_abc1",
+    "qual_ac1c2",
+    "qual_bc1c2",
+    "qual_abc1c2",
+)
+LONG_COUNTERS = (
+    "long_a",
+    "long_b",
+    "long_c",
+    "long_ab",
+    "long_c1c2",
+    "long_ac1",
+    "long_bc1",
+    "long_abc1",
+    "long_ac1c2",
+    "long_bc1c2",
+    "long_abc1c2",
+)
+TOTAL_COUNTERS = ("total_a", "total_b", "total_c", "fee_de_sent", "fee_de_recd")
+
+
+def create_dataset(packets: list[Packet]) -> xr.Dataset:
+    """Create dataset for a number of Hi Histogram packets.
+
+    Parameters
+    ----------
+    packets : list[Packet]
+        packet list
+
+    Returns
+    -------
+    xr.dataset
+        dataset with all metadata field data in xr.DataArray
+    """
+    dataset = allocate_histogram_dataset(len(packets))
+
+    # unpack the packets data into the Dataset
+    for i_epoch, packet in enumerate(packets):
+        dataset.epoch.data[i_epoch] = calc_start_time(
+            packet.data["CCSDS_MET"].raw_value
+        )
+        dataset.ccsds_met.data[i_epoch] = packet.data["CCSDS_MET"].raw_value
+        dataset.esa_step.data[i_epoch] = packet.data["ESA_STEP"].raw_value
+        dataset.acq_start.data[i_epoch] = packet.data["ACQ_START"].raw_value
+
+        # unpack 24 arrays of 90 12-bit unsigned integers
+        counters_binary_data = packet.data["COUNTERS"].raw_value
+        counter_ints = [
+            int(counters_binary_data[i * 12 : (i + 1) * 12], 2) for i in range(90 * 24)
+        ]
+        # populate the dataset with the unpacked integers
+        for i_counter, counter in enumerate(
+            (*QUALIFIED_COUNTERS, *LONG_COUNTERS, *TOTAL_COUNTERS)
+        ):
+            dataset[counter][i_epoch] = counter_ints[
```

Comment:
Above you used `.data[i_epoch]`, but here you use just `[i_epoch]` is there a difference between these setters?

Reply:
Good catch. I have no experience with xarray.DataArray. Both methods of setting values appear to do the same thing but I believe the correct/preferred way is without the `data` attribute. I'll make that change.

Reply:
Interestingly enough, if one tries to modify a coordinate variable without using the `data` attribute, xarray raises `TypeError: IndexVariable values cannot be modified`. Interestingly, using the `data` attribute has the desired effect with no error.

Reply:
Ahh, so the issue is with updating the coordinates only then? And I suppose it is because you are looping through this only updating one at a time instead of wanting to use a full `assign()` or `assign_coords()` that would reset the entire array. With the `.data` you are accessing the underlying numpy array and then mutating that without touching any of the metadata held within the xarray class.

Good to know, thanks for digging in!